File: 1/2021.03.31/503.Next Greater Element II/python/solution.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        res = [-1] * len(nums)
        stack = []
        for i in range(2 * len(nums)):
            while stack and nums[stack[-1]] < nums[i % len(nums)]:
                prev_idx = stack.pop()
                res[prev_idx] = nums[i % len(nums)]
            stack.append(i % len(nums))
        return res

File: 1/2021.03.31/503.Next Greater Element II/java/solution.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] res = new int[nums.length];
        Arrays.fill(res, -1);
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < nums.length * 2; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i % nums.length]) {
                res[stack.pop()] = nums[i % nums.length];
            }
            stack.push(i % nums.length);
        }
        return res;
    }
}
